components/dada-validate/src/validate/validator.rs

```diff
@@ -540,22 +541,49 @@ impl<'me> Validator<'me> {
     }
 }
 
-// Remove leading, trailing whitespace and common indentation from multiline strings.
+// Remove leading, trailing whitespace and common indent from multiline strings.
 fn convert_to_dada_string(s: &str) -> String {
+    // If the string has only one line, leave it and return immediately.
     if s.lines().count() == 1 {
         return s.to_string();
     }
-    let non_empty_lines = s
-        .lines()
-        .filter(|&line| !line.trim().is_empty())
-        .collect::<Vec<&str>>();
-
-    let indent = non_empty_lines
-        .iter()
-        .map(|line| line.chars().into_iter().take_while(|c| *c == ' ').count())
+
+    // Split string into lines and filter out empty lines.
+    let non_empty_line_iter = s.lines().filter(|&line| !line.trim().is_empty());
```

Comment:
A thought--

The common prefix has to be the prefix of *all* the lines. So maybe start with the first line and find its whitespace prefix. Then iterate over all the other lines. For each line, you compute the number of characters the two have in common and you find the minimum N of that. Then you keep the first N characters of the prefix.

In pseudocode...

```
fn count_characters_in_common(s1: &str, s2: &str) -> usize {
    s1.chars().zip(s2.chars()).take_while(|(c1, c2)| c1 == c2).count()
}

let non_empty_line_iter = s.lines().filter(|&line| !line.trim().is_empty());
let first_line = non_empty_line_iter.next();
let mut prefix = whitespace_prefix_of(first_line);
let shared_prefix_chars = non_empty_line_iter.map(|s| count_characters_in_common(prefix, s)).min();
prefix = &prefix[..shared_prefix_chars];
strip_prefix_from_original_string(prefix);
```

Reply:
This will make the code simpler and more efficient, thanks! I'll refactor it latter.

Reply:
One note is that the code I wrote counts characters, but for slicing, you really want to count utf-8 width. You can do that in various ways -- e.g. [`len_utf8`](https://doc.rust-lang.org/std/primitive.char.html#method.len_utf8) or the iterator (forgot the name) that gives you both characters and byte offsets.

Reply:
I modified the code to count characters to count bytes, I think this should be equivalent to `len_utf8`.